### PR TITLE
cp: implement backup support

### LINKED ISSUE #2231: cp: implement `--backup=[CONTROL]` support
[GNU man pages for `cp`](https://www.man7.org/linux/man-pages/man1/cp.1.html):

```
...
       --backup[=CONTROL]
              make a backup of each existing destination file
...
       The  backup  suffix is '~', unless set with --suffix or SIMPLE_BACKUP_SUFFIX.  The version control method
       may be selected via the --backup option or through the VERSION_CONTROL environment  variable.   Here  are
       the values:

       none, off
              never make backups (even if --backup is given)

       numbered, t
              make numbered backups

       existing, nil
              numbered if numbered backups exist, simple otherwise

       simple, never
              always make simple backups

       As a special case, cp makes a backup of SOURCE when the force and backup options are given and SOURCE and
       DEST are the same name for an existing, regular file.
```

The `cp` utility does not yet support `--backup=[CONTROL]`, but the [`mv`](https://github.com/uutils/coreutils/blob/master/src/uu/mv/src/mv.rs#L253) and `ln` utilities do.

Seems like this `--backup=[CONTROL]` support should be included in several utilities ([at least `cp`, `install`, `ln`, and `mv`](https://www.gnu.org/software/coreutils/manual/html_node/Backup-options.html)). It may be helpful to move the existing backup control code for `mv` and `ln` into the `uucore` or some shared library so it can be more easily maintained.

Specifically the [`BackupMode`](https://github.com/uutils/coreutils/blob/master/src/uu/mv/src/mv.rs#L44) enum is literally copied to [`ln`](https://github.com/uutils/coreutils/blob/master/src/uu/ln/src/ln.rs#L48), [`mv`](https://github.com/uutils/coreutils/blob/master/src/uu/mv/src/mv.rs#L44), and [`cp`](https://github.com/uutils/coreutils/blob/master/src/uu/cp/src/cp.rs#L173). But without the full implementation in `cp`.

Where should this code live in a shared space? `uucore/src/lib/features/backup_control.rs`?

### PR DESCRIPTION
**Pull request by blesson3**
This PR also includes a `uucore` implementation of backup control where it can live to be shared among the utilities needing backup control.

The backup control in `uucore` was sourced from `mv`'s existing backup control. In a follow-up PR (or this one on request) I can make changes to `mv` and other utilities so that they all use the `backup_control.rs` functionality in `uucore`.

fixes #2231

### CODE REVIEW

File: src/uucore/src/lib/mods/backup_control.rs
Comment (jhscheer) [inline src/uucore/src/lib/mods/backup_control.rs:43]: Please no URL shortening.

File: src/uucore/src/lib/mods/backup_control.rs
Comment (tertsdiepraam) [inline src/uucore/src/lib/mods/backup_control.rs:32]: `clap` can also handle environment variables and default values: https://docs.rs/clap/2.33.3/clap/struct.Arg.html#method.env

Although I can also understand if you implemented it here to ensure that the behaviour between `cp` and `mv` is the same.

File: tests/by-util/test_cp.rs
Comment (tertsdiepraam) [inline tests/by-util/test_cp.rs:337]: Great job on the tests!

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Great work! I just have some nitpicks about the argument handling and I agree that the url should not be shortened. I'm looking forward to the PR for `mv`!

File: src/uu/stdbuf/src/stdbuf.rs
Comment (blesson3) [reply src/uu/stdbuf/src/stdbuf.rs:28]: This change was the `$ cargo fmt` commit, which I had to have to pass the pre-commit script's `fmt` step. I'll remove this from the PR as it is completely unrelated.

File: src/uu/cp/src/cp.rs
Comment (blesson3) [reply src/uu/cp/src/cp.rs:371]: Great, I tried (but failed) to find functionality for clap that did exactly what `require_equals` does. I'll add that in plus a test to handle the `$ cp --backup file1 file2` case.

File: src/uucore/src/lib/mods/backup_control.rs
Comment (blesson3) [reply src/uucore/src/lib/mods/backup_control.rs:32]: Totally. I would rather have this functionality in `backup_control.rs` as it's easy to miss in other utilities implementing backups. Let me know if you require this change and I'll happily make it.

File: src/uucore/src/lib/mods/backup_control.rs
Comment (blesson3) [reply src/uucore/src/lib/mods/backup_control.rs]: Got it, fixed!

File: tests/by-util/test_cp.rs
Comment (blesson3) [reply tests/by-util/test_cp.rs:353]: Thanks!

File: src/uu/cp/src/cp.rs
Comment (blesson3) [reply src/uu/cp/src/cp.rs]: This is fixed and I added a test.

**Comment by blesson3** [comment]
I have an `mv` backup logic refactor PR ready to go. I'm going to wait until this is merged since the new `mv` PR ([branch](https://github.com/blesson3/coreutils/tree/mv-backup-refactor)) relies on `backup_control.rs`.

File: src/uu/cp/src/cp.rs
Comment (blesson3) [reply src/uu/cp/src/cp.rs]: Realized that I didn't merge the two flags, just pushed that change too.

File: src/uu/cp/src/cp.rs
Comment (tertsdiepraam) [inline src/uu/cp/src/cp.rs:597]: Ah I missed this on the previous review. You can use the `show_usage_error` here.

**Review by tertsdiepraam — COMMENTED** [review]
Looks good! I'll run the test suite to check if nothing breaks and then this can probably be merged.

File: src/uu/cp/src/cp.rs
Comment (blesson3) [reply src/uu/cp/src/cp.rs]: Got it, I moved this check to `uumain` so I could use `show_usage_error` and `return 1` instead of returning an error from `Options::from_matches`, which would result in weird output from the usage error and the returned error.

Comment (blesson3) [reply src/uu/cp/src/cp.rs]: I also added a test (`test_cp_backup_no_clobber_conflicting_options`) to check for the correct stderr from the `cp --backup --no-clobber` conflicting error.

**Pull request merged**
